File: kernel/io.c
#include "io.h"

void outb(uint16_t port, uint8_t val) {
	__asm__ volatile ("outb %b0, %w1" : : "a"(val), "Nd"(port) : "memory");
}

uint8_t inb(uint16_t port) {
	uint8_t res;
	__asm__ volatile ("inb %w1, %b0" : "=a"(res) : "Nd"(port) : "memory");
	return res;
}

void outw(uint16_t port, uint16_t val) {
	__asm__ volatile ("outw %w0, %w1" :: "a"(val), "Nd"(port) : "memory");
}

uint16_t inw(uint16_t port) {
	uint16_t res;
	__asm__ volatile ("inw %w1, %w0" : "=a"(res) : "Nd"(port) : "memory");
	return res;
}

void outl(uint16_t port, uint32_t val) {
	__asm__ volatile ("outl %k0, %w1" :: "a"(val), "Nd"(port) : "memory");
}

uint32_t inl(uint16_t port) {
	uint32_t res;
	__asm__ volatile ("inl %w1, %k0" : "=a"(res) : "Nd"(port) : "memory");
	return res;
}

void io_wait(void) {
	outb(0x80, 0);
}

File: src/zig/utils/io.zig
pub const Port = u16;

pub fn out(
    comptime bits: u16,
    port: Port,
    data: @Int(.unsigned, bits),
) void {
    switch (bits) {
        8 => out8(port, data),
        32 => out32(port, data),
        else => @compileError("unsupported I/O write bit width"),
    }
}

pub fn in(comptime bits: u16, port: Port) @Int(.unsigned, bits) {
    return switch (bits) {
        8 => in8(port),
        16 => in16(port),
        32 => in32(port),
        else => @compileError("unsupported I/O read bit width"),
    };
}

fn out8(port: Port, data: u8) void {
    asm volatile (
        \\ outb %%al, (%%dx)
        :
        : [data] "{al}" (data),
          [port] "{dx}" (port),
    );
}

fn in8(port: Port) u8 {
    return asm volatile (
        \\ inb (%%dx), %%al
        : [data] "={al}" (-> u8),
        : [port] "{dx}" (port),
    );
}

fn in16(port: Port) u16 {
    return asm volatile (
        \\ inw (%%dx), %%ax
        : [data] "={ax}" (-> u16),
        : [port] "{dx}" (port),
    );
}

fn out32(port: Port, data: u32) void {
    asm volatile (
        \\ outl %%eax, (%%dx)
        :
        : [data] "{eax}" (data),
          [port] "{dx}" (port),
    );
}

fn in32(port: Port) u32 {
    return asm volatile (
        \\ inl (%%dx), %%eax
        : [data] "={eax}" (-> u32),
        : [port] "{dx}" (port),
    );
}
